src/Nerdbank.MessagePack/Converters/ArrayConverter`1.cs

```diff
@@ -44,4 +50,75 @@ public override void Serialize(ref MessagePackWriter writer, ref TElement[]? val
 			elementConverter.Serialize(ref writer, ref value[i]!, context);
 		}
 	}
+
+	/// <inheritdoc/>
+	public override async ValueTask SerializeAsync(PipeWriter pipeWriter, TElement[]? value, SerializationContext context, CancellationToken cancellationToken)
+	{
+		Requires.NotNull(pipeWriter);
+
+		cancellationToken.ThrowIfCancellationRequested();
+
+		if (value is null)
+		{
+			WriteNil(pipeWriter);
+			return;
+		}
+
+		context.DepthStep();
+		if (elementConverter.PreferAsyncSerialization)
```

Comment:
I suspect it might be possible to avoid this virtual call if you can instead check it at folding time.

Reply:
Yes, I left a comment to that effect on this same line... :)